### ISSUE TITLE
Kasa-only Setup

### ISSUE THREAD

**Issue by paswai** [bentasker/tplink_to_influxdb#1]
Hi,
thank you for creating tplink_to_influxdb! I am only using Kasa smartplugs and don't own the newer Tapo ones. When running docker, a keyerror:

`File "/app/collect.py", line 80, in main
for tapo in config["tapo"]["devices"]: KeyError: 'tapo'`

In config.yml, I removed the Tapo part. If I include fake Tapo credentials and a fake Tapo device in config.yml:

`Failed to communicate with device washing-machine`

and two data points are written to InfluxDB. The docker container is being terminated and automatically restarted.

Is there a proper way to do it?
Best Regards,
Pascal

**Comment by bentasker** [comment]
Try putting this in your config

```
tapo:
    user:
    passw:
    devices:
kasa:
 your config
```

That should ensure the structure exists without generating unnecessary log noise.

You're right though, it would be better to be able to exclude it entirely - I'll look at doing that

**Comment by paswai** [comment]
Thanks for the quick answer! Your suggested solution worked and the error doesn't occur anymore. Do you think there is a semi-easy way to use a scheduler within docker or include a proper exit, thus, preventing the unexpected exit? (I don't have experience creating docker files)

**Comment by paswai** [comment]
Thank you for the detailed description - I will give it a try!
I am running the latest docker container on a Synology and while I don't see the error in the docker protocol, I get a message from Synology that the docker container was closed unexpectedly. I tried your code using the Synology terminal and here it worked correctly. Maybe, it is only a peculiar behaviour of the Synology...
Thanks again for your help!

**Comment by bentasker** [comment]
Hmm, I'm guessing Synology's setup is geared towards persistent containers (which, to be fair, is a more common usecase).

I've added a ticket to my todo list to look at adding support for running a persistent container. Thanks!

**Comment by bentasker** [comment]
Hey @paswai 

If you pull the latest image from docker hub you should now be able to add this to your config file
```yaml
poller:
    # If true, run in an infinite loop
    persist: true
    # Interval in seconds between runs
    interval: 20
```

Which'll result in the container not exiting unless it encounters errors (there's scope to improve handling of that, but this is a start)

**Comment by paswai** [comment]
Hi,
I pulled the latest image and added the new lines in the config file. The container protocol contains those new error lines:

```
Traceback (most recent call last):
  File "/app/collect.py", line 223, in <module>
    main()
  File "/app/collect.py", line 80, in main
    do_work(config, influxes)
  File "/app/collect.py", line 97, in do_work
    now_usage_w, today_usage = poll_kasa(kasa['ip'])
  File "/app/collect.py", line 162, in poll_kasa
    today_usage = p.emeter_today * 1000
TypeError: unsupported operand type(s) for *: 'NoneType' and 'int'
```

I might do a rollback to the old image today to see if the problem is the result of an (unintended) change on my end.

**Comment by bentasker** [comment]
Hey,

That's saying that  it didn't get a reading for today's usage from the socket, which should be unusual (though obviously could be handled better - https://projects.bentasker.co.uk/gils_projects/issue/utilities/tp-link-to-influxdb/6.html )

What model of smartplugs are you using?

Random guess: Have you blocked the plugs from accessing the internet at the router? I did that initially and found that the "today" readings don't work unless the plugs can reach a NTP server, so I've got a rule to specifically allow outbound UDP/123.

**Comment by paswai** [comment]
Hi,
I just tested an older version of your container (0.23) on my Synology and I get the same error. I will do some digging and give an update of the cause, but I think there is related to my setup (I changed the name_server on a Raspberry Pi which hosts my Pihole). I didn't (intentionally) block my SmartPlugs (Type KP115) from accessing the internet.
An update will be posted here, when I figured it out :-)

Anyway, thank you for quickly updating the container, it is really appreciated!
